### PR TITLE
[feat]Add Matrix Game 2.0 training

### PR DESCRIPTION
**Pull request by H1yori233**
## Added

- `examples/training/finetune/MatrixGame2.0/` folder contains relative scripts
- `fastvideo/pipelines/preprocess/matrixgame/matrixgame_preprocess_pipeline.py` preprocess pipeline
- `fastvideo/training/matrixgame_training_pipeline.py` training pipeline

Update foundation model [Matrix-Game-2.0-Foundation-Diffusers](https://huggingface.co/FastVideo/Matrix-Game-2.0-Foundation-Diffusers)

---

## Fixed
update `matrix_game` folder name to `matrixgame` to keep consistency.

---

## Changed
- register matrixgame
  - `fastvideo/pipelines/preprocess/v1_preprocess.py`
  - `fastvideo/dataset/dataloader/schema.py`
  - `fastvideo/configs/pipelines/wan.py`
- fix bug in ` fastvideo/models/dits/matrixgame/model.py` that return all zeros `encoder_hidden_states` for text embedding instead of `None`
- make pipeline works without text module
  - `fastvideo/pipelines/preprocess/preprocess_pipeline_base.py` 
  - `fastvideo/training/training_pipeline.py`
  - `fastvideo/models/dits/wanvideo.py`
- Expand restrictions on action cond dim.
  - `fastvideo/models/dits/matrixgame/action_module.py`
  - `fastvideo/models/dits/matrixgame/utils.py`

### CODE REVIEW

File: fastvideo/models/dits/matrixgame/causal_model.py [hunk 1]
```diff
@@ -1096,4 +1096,4 @@ def init_weights(self):
                     if block.action_model.proj_keyboard.bias is not None:
                         nn.init.zeros_(block.action_model.proj_keyboard.bias)
                 except AttributeError:
-                    pass
+                    pass
```
Comment (SolitaryThinker) [inline fastvideo/models/dits/matrixgame/causal_model.py]: revert

**Comment by H1yori233** [comment]
I think the CI fail is not related to my code?

**Comment by alexzms** [comment]
> I think the CI fail is not related to my code?

yeah please wait for some time, I will fix the CI

File: fastvideo/models/dits/matrixgame/action_module.py [hunk 2]
```diff
@@ -257,6 +257,15 @@ def forward(self, x, tt, th, tw, mouse_condition=None, keyboard_condition=None,
...
+        target_device = x.device
+        target_dtype = x.dtype
+        if mouse_condition is not None:
+            mouse_condition = mouse_condition.to(device=target_device,
+                                                 dtype=target_dtype)
+        if keyboard_condition is not None:
+            keyboard_condition = keyboard_condition.to(
+                device=target_device, dtype=target_dtype)
+
         B, N_frames, C = keyboard_condition.shape
```
Comment (alexzms) [inline fastvideo/models/dits/matrixgame/action_module.py:271]: if keyboard_condition is None, calling keyboard_condition.shape in line 269 will raise error.

File: fastvideo/models/dits/matrixgame/causal_model.py (hunk 1 above)
Comment (alexzms) [inline fastvideo/models/dits/matrixgame/causal_model.py:1099]: Please add tailing \n

File: fastvideo/training/matrixgame_training_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        """Override to properly handle I2V concatenation - call parent first, then concatenate image conditioning."""
+
+        # First, call parent method to prepare noise, timesteps, etc. for video latents
+        training_batch = super()._prepare_dit_inputs(training_batch)
+
+        assert isinstance(training_batch.image_latents, torch.Tensor)
+        image_latents = training_batch.image_latents.to(
+            get_local_torch_device(), dtype=torch.bfloat16)
+
+        temporal_compression_ratio = 4
```
Comment (alexzms) [inline fastvideo/training/matrixgame_training_pipeline.py]: Why hardcode? Could you change to reading from vae config just like in line 199?

File: fastvideo/training/matrixgame_training_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+            dim=1)
+
+        return training_batch
+
+    def _build_input_kwargs(self,
+                            training_batch: TrainingBatch) -> TrainingBatch:
+
+        # Image Embeds for conditioning
+        image_embeds = training_batch.image_embeds
+        assert torch.isnan(image_embeds).sum() == 0
```
Comment (alexzms) [inline fastvideo/training/matrixgame_training_pipeline.py:154]: Just asking, will this affect training performance? Maybe call it in lower frequency/only call in debug mode is better?

File: fastvideo/training/matrixgame_training_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        assert self.seed is not None
+        sampling_param.seed = self.seed
+
+        latents_size = [(sampling_param.num_frames - 1) // 4 + 1,
+                        sampling_param.height // 8, sampling_param.width // 8]
+        n_tokens = latents_size[0] * latents_size[1] * latents_size[2]
+        temporal_compression_factor = training_args.pipeline_config.vae_config.arch_config.temporal_compression_ratio
+        num_frames = (training_args.num_latent_t -
+                      1) * temporal_compression_factor + 1
+        sampling_param.num_frames = num_frames
```
Comment (alexzms) [inline fastvideo/training/matrixgame_training_pipeline.py:202]: In line 200, we updated num_frames without changing n_tokens, is this behavior expected?

File: fastvideo/training/matrixgame_training_pipeline.py (hunk 4 above)
Comment (H1yori233) [reply fastvideo/training/matrixgame_training_pipeline.py:154]: This is same as [wan_i2v_training_pipeline.py](https://github.com/hao-ai-lab/FastVideo/blob/main/fastvideo/training/wan_i2v_training_pipeline.py), so it should be okay I think

File: fastvideo/training/matrixgame_training_pipeline.py (hunk 5 above)
Comment (H1yori233) [reply fastvideo/training/matrixgame_training_pipeline.py:202]: This is same as [wan_i2v_training_pipeline.py](https://github.com/hao-ai-lab/FastVideo/blob/main/fastvideo/training/wan_i2v_training_pipeline.py), so it should be okay I think

File: fastvideo/models/dits/matrixgame/causal_model.py (hunk 1 above)
Comment (H1yori233) [reply fastvideo/models/dits/matrixgame/causal_model.py:1099]: I noticed that most of the existing files don't have a trailing newline, so this should be expected?
and I tried to copy-paste the origin code, but my Git isn't showing any modifications, so this might be a GitHub display bug.🤔

File: fastvideo/models/dits/matrixgame/action_module.py (hunk 2 above)
Comment (H1yori233) [reply fastvideo/models/dits/matrixgame/action_module.py:271]: Now if `keyboard_condition` is None, it will directly return the `x`, since keyboard_cond is always required in MatrixGame2.0.

Comment (H1yori233) [reply fastvideo/models/dits/matrixgame/action_module.py:271]:
```
        if keyboard_condition is not None:
            keyboard_condition = keyboard_condition.to(
                device=target_device, dtype=target_dtype)
        else:
            return x
```

**Comment by H1yori233** [comment]
Thanks for reviewing! Now all issues are solved I think.

**Comment by alexzms** [comment]
Cool, after the CI passing, I think we can merge this PR.

**Comment by H1yori233** [comment]
vsa test failed again, is this related to my code?
and other tests all passed

**Comment by alexzms** [comment]
Please wait. Currently solving this problem

**Comment by alexzms** [comment]
Please rebase the latest main branch

**Comment by H1yori233** [comment]
The test results are inconsistent, it passes once and fails the next time. it’s a ci-related issue. I feels like a waste of time to keep troubleshooting this. 🤔

**Review by alexzms — APPROVED** [review]
The code looks great to me. I think we can merge this code. The CI problem is quite strange
